File: site/js/engine.js
/* Motore di TRANS in JavaScript — porto fedele di trans/cards.py, trans/rules.py
 * e trans/engine.py.
 *
 * Serve alla versione peer-to-peer: su GitHub Pages non gira nessun processo,
 * quindi le regole devono stare nel browser di chi apre la stanza.
 *
 * Le carte sono stringhe ("AH", "10S", "2C"): stesso formato che il server
 * Python manda sul filo, cosi' la UI non distingue da dove arriva lo stato.
 *
 * L'unica fonte di verita' sulle regole restano i test Python. site/tests/
 * riesegue qui le partite registrate da quelli e pretende gli stessi punti.
 */

export const SUITS = ["H", "D", "C", "S"];
export const TRUMP_SUIT = "H";

const RANK_LABELS = { 11: "J", 12: "Q", 13: "K", 14: "A" };
const LABEL_RANKS = { J: 11, Q: 12, K: 13, A: 14 };
export const RANKS = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14];

// Ordine di visualizzazione: briscola per prima.
const DISPLAY_ORDER = ["H", "S", "D", "C"];

export const SUIT_NAMES = { H: "cuori", D: "quadri", C: "fiori", S: "picche" };

export function rankOf(code) {
  const label = code.slice(0, -1);
  return LABEL_RANKS[label] !== undefined ? LABEL_RANKS[label] : parseInt(label, 10);
}

export function suitOf(code) {
  return code.slice(-1);
}

export function makeCode(rank, suit) {
  return (RANK_LABELS[rank] || String(rank)) + suit;
}

export function fullDeck() {
  const deck = [];
  for (const suit of DISPLAY_ORDER) {
    for (const rank of RANKS) deck.push(makeCode(rank, suit));
  }
  return deck;
}

/** Chiave d'ordinamento in una presa: briscola > seme di uscita > scarto. */
export function trickKey(code, leadSuit, trump) {
  const suit = suitOf(code);
  const tier = trump && suit === trump ? 2 : suit === leadSuit ? 1 : 0;
  return tier * 100 + rankOf(code);
}

export function beats(card, other, leadSuit, trump) {
  return trickKey(card, leadSuit, trump) > trickKey(other, leadSuit, trump);
}

/** Indice del giocatore che vince una presa completa. */
export function trickWinner(plays, trump) {
  if (!plays.length) throw new Error("presa vuota");
  const lead = suitOf(plays[0].card);
  let best = plays[0];
  for (const play of plays.slice(1)) {
    if (beats(play.card, best.card, lead, trump)) best = play;
  }
  return best.player;
}

export function sortHand(codes) {
  return [...codes].sort((a, b) => {
    const bySuit = DISPLAY_ORDER.indexOf(suitOf(a)) - DISPLAY_ORDER.indexOf(suitOf(b));
    return bySuit !== 0 ? bySuit : rankOf(b) - rankOf(a);
  });
}

export function shuffle(items, random = Math.random) {
  const out = [...items];
  for (let i = out.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1));
    [out[i], out[j]] = [out[j], out[i]];
  }
  return out;
}

export function deal(nPlayers, cardsEach, random = Math.random) {
  if (nPlayers * cardsEach > 52) {
    throw new Error(`servono ${nPlayers * cardsEach} carte ma il mazzo ne ha 52`);
  }
  const deck = shuffle(fullDeck(), random);
  const hands = [];
  for (let i = 0; i < nPlayers; i++) {
    hands.push(sortHand(deck.slice(i * cardsEach, (i + 1) * cardsEach)));
  }
  return hands;
}

/* ------------------------------------------------------------------ regole */

export const RoundKind = {
  NORMAL: "normal",
  NO_TRUMP: "no_trump",
  BLIND: "blind",
  MISERE: "misere",
};

const KIND_LABELS = {
  normal: "Normale",
  no_trump: "NO BRISCOLA",
  blind: "BUIO",
  misere: "A PERDERE",
};

export const SPECIAL_ROUND_CARDS = 7;
export const MAX_NORMAL_CARDS = 7;
const SPECIAL_SEQUENCE = [RoundKind.NO_TRUMP, RoundKind.BLIND, RoundKind.MISERE];

export const HIT_BASE = 10;
export const HIT_PER_TRICK = 5;
export const MISS_PER_TRICK = 1;
export const MISERE_PER_TRICK = -5;

export const MIN_PLAYERS = 2;
export const MAX_PLAYERS = 6;

function roundSpec(number, cards, kind) {
  return {
    number,
    cards,
    kind,
    kindLabel: KIND_LABELS[kind],
    trump: kind === RoundKind.NO_TRUMP ? null : TRUMP_SUIT,
    hasBidding: kind !== RoundKind.MISERE,
    blindBidding: kind === RoundKind.BLIND,
    title: kind === RoundKind.NORMAL ? `${cards} carte` : `${KIND_LABELS[kind]} · ${cards} carte`,
  };
}

/**
 * Veloce (10): 7,6,5,4,3,2,1 + NO BRISCOLA, BUIO, A PERDERE.
 * Lunga  (20): come sopra, poi la risalita 1..7 e di nuovo i tre speciali.
 */
export function buildSchedule(mode) {
  const plan = [];
  for (let n = MAX_NORMAL_CARDS; n >= 1; n--) plan.push([n, RoundKind.NORMAL]);
  for (const kind of SPECIAL_SEQUENCE) plan.push([SPECIAL_ROUND_CARDS, kind]);
  if (mode === "long") {
    for (let n = 1; n <= MAX_NORMAL_CARDS; n++) plan.push([n, RoundKind.NORMAL]);
    for (const kind of SPECIAL_SEQUENCE) plan.push([SPECIAL_ROUND_CARDS, kind]);
  }
  return plan.map(([cards, kind], i) => roundSpec(i + 1, cards, kind));
}

export const MODE_LABELS = {
  fast: "Partita veloce (10 round)",
  long: "Partita lunga (20 round)",
};

/**
 * A PERDERE: -5 per presa. Dichiarazione centrata: 10 + 5 per presa dichiarata.
 * Dichiarazione sbagliata: 1 punto per presa fatta.
 */
export function scoreRound(kind, bid, tricks) {
  if (kind === RoundKind.MISERE) return MISERE_PER_TRICK * tricks;
  if (bid !== null && bid !== undefined && bid === tricks) return HIT_BASE + HIT_PER_TRICK * bid;
  return MISS_PER_TRICK * tricks;
}

/* ------------------------------------------------------------------ partita */

export const Phase = {
  BIDDING: "bidding",
  PLAYING: "playing",
  ROUND_OVER: "round_over",
  GAME_OVER: "game_over",
};

export class IllegalMove extends Error {}

export class Game {
  /**
   * @param {Array} players  {id, name, isBot, botLevel}
   * @param {Object} options  mode, dealer, forbidExactTotal, random, dealFn
   */
  constructor(players, options = {}) {
    if (players.length < MIN_PLAYERS || players.length > MAX_PLAYERS) {
      throw new Error(`servono da ${MIN_PLAYERS} a ${MAX_PLAYERS} giocatori`);
    }
    if (new Set(players.map((p) => p.id)).size !== players.length) {
      throw new Error("id giocatori duplicati");
    }

    this.players = players.map((p) => ({
      id: p.id,
      name: p.name,
      isBot: !!p.isBot,
      botLevel: p.botLevel || "normale",
      score: 0,
      hand: [],
      bid: null,
      tricks: 0,
    }));
    this.mode = options.mode || "fast";
    this.schedule = buildSchedule(this.mode);
    this.random = options.random || Math.random;
    // dealFn esiste per i test: permette di rigiocare mani gia' distribuite.
    this.dealFn = options.dealFn || ((n, cards) => deal(n, cards, this.random));
    // Regola dell'impiccato, attiva come nel motore Python.
    this.forbidExactTotal = options.forbidExactTotal !== false;

    this.roundIndex = 0;
    this.dealer = (options.dealer || 0) % this.players.length;
    this.turn = 0;
    this.phase = Phase.BIDDING;
    this.currentTrick = [];
    this.lastTrick = null;
    this.playedCards = [];
    this.results = [];
    this.log = [];

    this._startRound();
  }

  get n() {
    return this.players.length;
  }

  get spec() {
    return this.schedule[this.roundIndex];
  }

  get trump() {
    return this.spec.trump;
  }

  get isOver() {
    return this.phase === Phase.GAME_OVER;
  }

  get leadSuit() {
    return this.currentTrick.length ? suitOf(this.currentTrick[0].card) : null;
  }

  indexOf(playerId) {
    const i = this.players.findIndex((p) => p.id === playerId);
    if (i < 0) throw new Error(`giocatore sconosciuto: ${playerId}`);
    return i;
  }

  currentActor() {
    return this.phase === Phase.BIDDING || this.phase === Phase.PLAYING ? this.turn : null;
  }

  handVisibleToOwner() {
    return !(this.spec.blindBidding && this.phase === Phase.BIDDING);
  }

  _log(message) {
    this.log.push(message);
    if (this.log.length > 200) this.log = this.log.slice(-200);
  }

  _startRound() {
    const spec = this.spec;
    const hands = this.dealFn(this.n, spec.cards);
    this.players.forEach((player, i) => {
      player.hand = [...hands[i]];
      player.bid = null;
      player.tricks = 0;
    });
    this.currentTrick = [];
    this.lastTrick = null;
    this.playedCards = [];
    this.turn = (this.dealer + 1) % this.n;
    this.phase = spec.hasBidding ? Phase.BIDDING : Phase.PLAYING;
    this._log(`Round ${spec.number}/${this.schedule.length} — ${spec.title}`);
    if (!spec.hasBidding) this._log("A PERDERE: nessuna scommessa, ogni presa vale -5.");
  }

  /* ---------------------------------------------------------- dichiarazioni */

  /** La dichiarazione vietata a chi chiude il giro: quella che pareggia il conto. */
  forbiddenBid(playerIndex) {
    if (!this.forbidExactTotal || playerIndex !== this.dealer) return null;
    const declared = this.players.reduce((sum, p) => sum + (p.bid === null ? 0 : p.bid), 0);
    const forbidden = this.spec.cards - declared;
    return forbidden >= 0 && forbidden <= this.spec.cards ? forbidden : null;
  }

  legalBids(playerIndex) {
    if (this.phase !== Phase.BIDDING || playerIndex !== this.turn) return [];
    const forbidden = this.forbiddenBid(playerIndex);
    const options = [];
    for (let b = 0; b <= this.spec.cards; b++) if (b !== forbidden) options.push(b);
    return options;
  }

  placeBid(playerId, value) {
    const idx = this.indexOf(playerId);
    if (this.phase !== Phase.BIDDING) throw new IllegalMove("non e' il momento di scommettere");
    if (idx !== this.turn) throw new IllegalMove("non e' il tuo turno di scommettere");
    if (value === this.forbiddenBid(idx)) {
      throw new IllegalMove(
        `chiudi tu il giro: con ${value} la somma farebbe esattamente ${this.spec.cards}, e non e' permesso`
      );
    }
    if (!this.legalBids(idx).includes(value)) {
      throw new IllegalMove(`scommessa non valida: ${value}`);
    }

    const player = this.players[idx];
    player.bid = value;
    this._log(`${player.name} scommette ${value}.`);

    if (this.players.every((p) => p.bid !== null)) {
      this.phase = Phase.PLAYING;
      this.turn = (this.dealer + 1) % this.n;
      const total = this.players.reduce((s, p) => s + p.bid, 0);
      this._log(`Scommesse chiuse: ${total} prese dichiarate su ${this.spec.cards}.`);
    } else {
      this.turn = (this.turn + 1) % this.n;
    }
  }

  /* ---------------------------------------------------------------- gioco */

  /** Obbligo di rispondere al seme; se non ce l'hai, giochi quello che vuoi. */
  legalCards(playerIndex) {
    if (this.phase !== Phase.PLAYING || playerIndex !== this.turn) return [];
    const hand = this.players[playerIndex].hand;
    const lead = this.leadSuit;
    if (!lead) return [...hand];
    const sameSuit = hand.filter((c) => suitOf(c) === lead);
    return sameSuit.length ? sameSuit : [...hand];
  }

  playCard(playerId, card) {
    const idx = this.indexOf(playerId);
    if (this.phase !== Phase.PLAYING) throw new IllegalMove("non e' il momento di giocare una carta");
    if (idx !== this.turn) throw new IllegalMove("non e' il tuo turno");
    const player = this.players[idx];
    if (!player.hand.includes(card)) throw new IllegalMove(`non hai ${card} in mano`);
    if (!this.legalCards(idx).includes(card)) {
      const lead = this.leadSuit;
      throw new IllegalMove(lead ? `devi rispondere a ${SUIT_NAMES[lead]}` : "carta non giocabile");
    }

    player.hand.splice(player.hand.indexOf(card), 1);
    this.currentTrick.push({ player: idx, card });
    this.playedCards.push(card);
    this._log(`${player.name} gioca ${card}.`);

    if (this.currentTrick.length === this.n) this._resolveTrick();
    else this.turn = (this.turn + 1) % this.n;
  }

  _resolveTrick() {
    const winner = trickWinner(this.currentTrick, this.trump);
    this.players[winner].tricks += 1;
    this.lastTrick = { plays: [...this.currentTrick], winner };
    this._log(`Presa a ${this.players[winner].name}.`);
    this.currentTrick = [];
    this.turn = winner;
    if (!this.players.some((p) => p.hand.length)) this._endRound();
  }

  _endRound() {
    const spec = this.spec;
    const rows = this.players.map((player) => {
      const delta = scoreRound(spec.kind, player.bid, player.tricks);
      player.score += delta;
      return {
        player_id: player.id,
        name: player.name,
        bid: player.bid,
        tricks: player.tricks,
        delta,
        total: player.score,
      };
    });
    this.results.push({ roundNumber: spec.number, kind: spec.kind, rows });
    this.phase = Phase.ROUND_OVER;
    const summary = rows.map((r) => `${r.name} ${r.delta >= 0 ? "+" : ""}${r.delta}`).join(", ");
    this._log(`Fine round ${spec.number}: ${summary}.`);
  }

  advanceRound() {
    if (this.phase !== Phase.ROUND_OVER) throw new IllegalMove("il round non e' finito");
    if (this.roundIndex + 1 >= this.schedule.length) {
      this.phase = Phase.GAME_OVER;
      this._log(`Partita finita. Vince ${this.winnerNames()}.`);
      return;
    }
    this.roundIndex += 1;
    this.dealer = (this.dealer + 1) % this.n;
    this._startRound();
  }

  standings() {
    return [...this.players].sort((a, b) => b.score - a.score);
  }

  winnerNames() {
    const best = Math.max(...this.players.map((p) => p.score));
    const names = this.players.filter((p) => p.score === best).map((p) => p.name);
    return `${names.join(" e ")} (${best} punti)`;
  }

  /* -------------------------------------------------------------- snapshot */

  /** Stesso formato del server Python: la UI non sa da dove arriva. */
  snapshot(viewerId) {
    let viewer = null;
    if (viewerId) {
      const i = this.players.findIndex((p) => p.id === viewerId);
      if (i >= 0) viewer = i;
    }
    const spec = this.spec;
    const actor = this.currentActor();

    let hand = [];
    let legalCards = [];
    let legalBids = [];
    let forbiddenBid = null;
    if (viewer !== null) {
      if (this.handVisibleToOwner()) hand = sortHand(this.players[viewer].hand);
      legalCards = this.legalCards(viewer);
      legalBids = this.legalBids(viewer);
      forbiddenBid = this.forbiddenBid(viewer);
    }

    return {
      phase: this.phase,
      mode: this.mode,
      round: {
        number: spec.number,
        total: this.schedule.length,
        cards: spec.cards,
        kind: spec.kind,
        kind_label: spec.kindLabel,
        title: spec.title,
        trump: spec.trump,
        blind: spec.blindBidding,
        has_bidding: spec.hasBidding,
      },
      players: this.players.map((p, i) => ({
        id: p.id,
        name: p.name,
        is_bot: p.isBot,
        bot_level: p.isBot ? p.botLevel : null,
        score: p.score,
        bid: p.bid,
        tricks: p.tricks,
        cards_left: p.hand.length,
        is_dealer: i === this.dealer,
        is_turn: actor === i,
      })),
      you: viewer,
      hand,
      hand_hidden: viewer !== null && !this.handVisibleToOwner(),
      legal_cards: legalCards,
      legal_bids: legalBids,
      forbidden_bid: forbiddenBid,
      trick: this.currentTrick.map((p) => ({ player: p.player, card: p.card })),
      lead_suit: this.leadSuit,
      played: [...this.playedCards],
      last_trick: this.lastTrick,
      last_result: this.results.length
        ? {
            round: this.results[this.results.length - 1].roundNumber,
            rows: this.results[this.results.length - 1].rows,
          }
        : null,
      standings: this.standings().map((p) => ({ id: p.id, name: p.name, score: p.score })),
      schedule: this.schedule.map((s) => ({
        number: s.number,
        cards: s.cards,
        kind: s.kind,
        title: s.title,
      })),
      round_index: this.roundIndex,
      log: this.log.slice(-40),
      winner: this.isOver ? this.winnerNames() : null,
    };
  }
}

File: trans/engine.py
"""Macchina a stati di una partita di TRANS.

L'engine e' puro: nessuna dipendenza da rete o UI, e deterministico a parita'
di seed. Il server ci costruisce sopra il multiplayer, i bot ci girano contro.
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from enum import Enum

from .cards import Card, Suit, deal, sort_hand, trick_winner
from .rules import GameMode, RoundKind, RoundSpec, build_schedule, max_players_for, score_round

MIN_PLAYERS = 2
MAX_PLAYERS = 6


class Phase(Enum):
    BIDDING = "bidding"
    PLAYING = "playing"
    ROUND_OVER = "round_over"
    GAME_OVER = "game_over"


class IllegalMove(Exception):
    """Mossa rifiutata dalle regole (turno sbagliato, carta non giocabile, ...)."""


@dataclass
class Player:
    id: str
    name: str
    is_bot: bool = False
    bot_level: str = "normale"
    score: int = 0
    hand: list[Card] = field(default_factory=list)
    bid: int | None = None
    tricks: int = 0


@dataclass
class RoundResult:
    round_number: int
    kind: RoundKind
    rows: list[dict]  # {player_id, name, bid, tricks, delta, total}


class Game:
    def __init__(
        self,
        players: list[Player],
        mode: GameMode = GameMode.FAST,
        seed: int | None = None,
        dealer: int = 0,
        forbid_exact_total: bool = True,
    ) -> None:
        if not MIN_PLAYERS <= len(players) <= MAX_PLAYERS:
            raise ValueError(f"servono da {MIN_PLAYERS} a {MAX_PLAYERS} giocatori")
        if len(players) > max_players_for(7):
            raise ValueError("troppi giocatori per un mazzo da 52 carte")
        if len({p.id for p in players}) != len(players):
            raise ValueError("id giocatori duplicati")

        self.players = players
        self.mode = mode
        self.schedule: list[RoundSpec] = build_schedule(mode)
        self.rng = random.Random(seed)
        self.seed = seed
        # Regola dell'impiccato: l'ultimo a parlare non puo' pareggiare il conto,
        # cosi' almeno un giocatore sbaglia per forza la dichiarazione.
        self.forbid_exact_total = forbid_exact_total

        self.round_index = 0
        self.dealer = dealer % len(players)
        self.turn = 0
        self.phase = Phase.BIDDING
        self.current_trick: list[tuple[int, Card]] = []
        self.last_trick: dict | None = None
        self.played_cards: list[Card] = []
        self.results: list[RoundResult] = []
        self.log: list[str] = []

        self._start_round()

    # ------------------------------------------------------------------ info

    @property
    def n(self) -> int:
        return len(self.players)

    @property
    def spec(self) -> RoundSpec:
        return self.schedule[self.round_index]

    @property
    def trump(self) -> Suit | None:
        return self.spec.trump

    @property
    def is_over(self) -> bool:
        return self.phase is Phase.GAME_OVER

    def index_of(self, player_id: str) -> int:
        for i, p in enumerate(self.players):
            if p.id == player_id:
                return i
        raise KeyError(player_id)

    def current_actor(self) -> int | None:
        """Indice del giocatore che deve muovere, o None se non tocca a nessuno."""
        if self.phase in (Phase.BIDDING, Phase.PLAYING):
            return self.turn
        return None

    def hand_visible_to_owner(self) -> bool:
        """Nel BUIO le proprie carte restano coperte finche' non si e' scommesso."""
        return not (self.spec.blind_bidding and self.phase is Phase.BIDDING)

    # ------------------------------------------------------------------ setup

    def _start_round(self) -> None:
        spec = self.spec
        hands = deal(self.n, spec.cards, self.rng)
        for player, hand in zip(self.players, hands):
            player.hand = hand
            player.bid = None
            player.tricks = 0
        self.current_trick = []
        self.last_trick = None
        self.played_cards = []
        self.turn = (self.dealer + 1) % self.n
        self.phase = Phase.BIDDING if spec.has_bidding else Phase.PLAYING
        self._log(f"Round {spec.number}/{len(self.schedule)} — {spec.title}")
        if not spec.has_bidding:
            self._log("A PERDERE: nessuna scommessa, ogni presa vale -5.")

    def _log(self, message: str) -> None:
        self.log.append(message)
        del self.log[:-200]

    # -------------------------------------------------------------- scommesse

    def legal_bids(self, player_index: int) -> list[int]:
        if self.phase is not Phase.BIDDING or player_index != self.turn:
            return []
        forbidden = self.forbidden_bid(player_index)
        return [b for b in range(self.spec.cards + 1) if b != forbidden]

    def forbidden_bid(self, player_index: int) -> int | None:
        """La dichiarazione vietata all'ultimo a parlare, se ce n'e' una.

        La somma delle dichiarazioni non puo' fare esattamente il numero di
        prese in palio: chi chiude il giro deve sbilanciare il conto.
        """
        if not self.forbid_exact_total or not self._is_last_bidder(player_index):
            return None
        declared = sum(p.bid for p in self.players if p.bid is not None)
        forbidden = self.spec.cards - declared
        return forbidden if 0 <= forbidden <= self.spec.cards else None

    def _is_last_bidder(self, player_index: int) -> bool:
        """Si parla da sinistra del mazziere, quindi il mazziere chiude."""
        return player_index == self.dealer

    def place_bid(self, player_id: str, value: int) -> None:
        idx = self.index_of(player_id)
        if self.phase is not Phase.BIDDING:
            raise IllegalMove("non e' il momento di scommettere")
        if idx != self.turn:
            raise IllegalMove("non e' il tuo turno di scommettere")
        if value == self.forbidden_bid(idx):
            raise IllegalMove(
                f"chiudi tu il giro: con {value} la somma farebbe esattamente "
                f"{self.spec.cards}, e non e' permesso"
            )
        if value not in self.legal_bids(idx):
            raise IllegalMove(f"scommessa non valida: {value}")

        player = self.players[idx]
        player.bid = value
        self._log(f"{player.name} scommette {value}.")

        if all(p.bid is not None for p in self.players):
            self.phase = Phase.PLAYING
            self.turn = (self.dealer + 1) % self.n
            total = sum(p.bid or 0 for p in self.players)
            self._log(f"Scommesse chiuse: {total} prese dichiarate su {self.spec.cards}.")
        else:
            self.turn = (self.turn + 1) % self.n

    # ------------------------------------------------------------------ gioco

    @property
    def lead_suit(self) -> Suit | None:
        return self.current_trick[0][1].suit if self.current_trick else None

    def legal_cards(self, player_index: int) -> list[Card]:
        """Obbligo di rispondere al seme; se non lo hai, giochi quello che vuoi."""
        if self.phase is not Phase.PLAYING or player_index != self.turn:
            return []
        hand = self.players[player_index].hand
        lead = self.lead_suit
        if lead is None:
            return list(hand)
        same_suit = [c for c in hand if c.suit is lead]
        return same_suit or list(hand)

    def play_card(self, player_id: str, card: Card) -> None:
        idx = self.index_of(player_id)
        if self.phase is not Phase.PLAYING:
            raise IllegalMove("non e' il momento di giocare una carta")
        if idx != self.turn:
            raise IllegalMove("non e' il tuo turno")
        player = self.players[idx]
        if card not in player.hand:
            raise IllegalMove(f"non hai {card} in mano")
        if card not in self.legal_cards(idx):
            lead = self.lead_suit
            raise IllegalMove(
                f"devi rispondere a {lead.italian}" if lead else "carta non giocabile"
            )

        player.hand.remove(card)
        self.current_trick.append((idx, card))
        self.played_cards.append(card)
        self._log(f"{player.name} gioca {card}.")

        if len(self.current_trick) == self.n:
            self._resolve_trick()
        else:
            self.turn = (self.turn + 1) % self.n

    def _resolve_trick(self) -> None:
        winner = trick_winner(self.current_trick, self.trump)
        self.players[winner].tricks += 1
        self.last_trick = {
            "plays": [{"player": i, "card": c.code} for i, c in self.current_trick],
            "winner": winner,
        }
        self._log(f"Presa a {self.players[winner].name}.")
        self.current_trick = []
        self.turn = winner
        if not any(p.hand for p in self.players):
            self._end_round()

    # ---------------------------------------------------------- fine round

    def _end_round(self) -> None:
        spec = self.spec
        rows = []
        for player in self.players:
            delta = score_round(spec.kind, player.bid, player.tricks, spec.cards)
            player.score += delta
            rows.append(
                {
                    "player_id": player.id,
                    "name": player.name,
                    "bid": player.bid,
                    "tricks": player.tricks,
                    "delta": delta,
                    "total": player.score,
                }
            )
        self.results.append(RoundResult(spec.number, spec.kind, rows))
        self.phase = Phase.ROUND_OVER
        summary = ", ".join(f"{r['name']} {r['delta']:+d}" for r in rows)
        self._log(f"Fine round {spec.number}: {summary}.")

    def advance_round(self) -> None:
        """Passa al round successivo (o chiude la partita)."""
        if self.phase is not Phase.ROUND_OVER:
            raise IllegalMove("il round non e' finito")
        if self.round_index + 1 >= len(self.schedule):
            self.phase = Phase.GAME_OVER
            self._log(f"Partita finita. Vince {self.winner_names()}.")
            return
        self.round_index += 1
        self.dealer = (self.dealer + 1) % self.n
        self._start_round()

    def standings(self) -> list[Player]:
        return sorted(self.players, key=lambda p: -p.score)

    def winner_names(self) -> str:
        best = max(p.score for p in self.players)
        names = [p.name for p in self.players if p.score == best]
        return " e ".join(names) + f" ({best} punti)"

    # ------------------------------------------------------------ snapshot

    def snapshot(self, viewer_id: str | None = None) -> dict:
        """Stato serializzabile, con la mano visibile solo al proprietario."""
        viewer = None
        if viewer_id is not None:
            try:
                viewer = self.index_of(viewer_id)
            except KeyError:
                viewer = None

        spec = self.spec
        players = [
            {
                "id": p.id,
                "name": p.name,
                "is_bot": p.is_bot,
                "bot_level": p.bot_level if p.is_bot else None,
                "score": p.score,
                "bid": p.bid,
                "tricks": p.tricks,
                "cards_left": len(p.hand),
                "is_dealer": i == self.dealer,
                "is_turn": self.current_actor() == i,
            }
            for i, p in enumerate(self.players)
        ]

        hand: list[str] = []
        legal_cards: list[str] = []
        legal_bids: list[int] = []
        forbidden_bid: int | None = None
        if viewer is not None:
            if self.hand_visible_to_owner():
                hand = [c.code for c in sort_hand(self.players[viewer].hand)]
            legal_cards = [c.code for c in self.legal_cards(viewer)]
            legal_bids = self.legal_bids(viewer)
            forbidden_bid = self.forbidden_bid(viewer)

        return {
            "phase": self.phase.value,
            "mode": self.mode.value,
            "round": {
                "number": spec.number,
                "total": len(self.schedule),
                "cards": spec.cards,
                "kind": spec.kind.value,
                "kind_label": spec.kind.label,
                "title": spec.title,
                "trump": spec.trump.value if spec.trump else None,
                "blind": spec.blind_bidding,
                "has_bidding": spec.has_bidding,
            },
            "players": players,
            "you": viewer,
            "hand": hand,
            "hand_hidden": viewer is not None and not self.hand_visible_to_owner(),
            "legal_cards": legal_cards,
            "legal_bids": legal_bids,
            "forbidden_bid": forbidden_bid,
            "trick": [{"player": i, "card": c.code} for i, c in self.current_trick],
            "lead_suit": self.lead_suit.value if self.lead_suit else None,
            "played": [c.code for c in self.played_cards],
            "last_trick": self.last_trick,
            "last_result": (
                {"round": self.results[-1].round_number, "rows": self.results[-1].rows}
                if self.results
                else None
            ),
            "standings": [{"id": p.id, "name": p.name, "score": p.score} for p in self.standings()],
            "schedule": [
                {"number": s.number, "cards": s.cards, "kind": s.kind.value, "title": s.title}
                for s in self.schedule
            ],
            "round_index": self.round_index,
            "log": self.log[-40:],
            "winner": self.winner_names() if self.is_over else None,
        }
